helm: split proxy/auth documentation (#19881)
gravitational/teleport: docs/pages/deploy-a-cluster/helm-deployments/custom.mdx

description: Install and configure a Teleport cluster with a custom configuration using Helm
---

In this guide, we'll explain how to set up a Teleport cluster in Kubernetes
using a custom [`teleport.yaml`](../../reference/config.mdx) config file
using Teleport Helm charts.

This setup can be useful when you already have an existing Teleport cluster and would
like to start running it in Kubernetes, or when migrating your setup from a legacy
version of the Helm charts.

<Admonition type="warning">
Those instructions are both for v12 Teleport and v12 `teleport-cluster` chart.

> If you are running an older Teleport version, use the version selector to consult

Reviewer: ```suggestion
If you are running an older Teleport version, use the version selector to choose
```

  [Information on supported Teleport storage backends](../../reference/backends.mdx)

  Manually configuring NFS-based storage or `ReadWriteMany` volume claims is **NOT** supported for an HA deployment and will result in errors.
</Admonition>

Write the following `my-values.yaml` file, and adapt the teleport configuration as you want.
You can find all possible configuration fields in the [Teleport Config Reference](../../reference/config.mdx).

```yaml
chartMode: scratch

auth:
  teleportConfig:
    # put your teleport.yaml auth configuration here
    teleport:
      log:
        output: stderr
        severity: INFO

    auth_service:
      enabled: true
      listen_addr: 0.0.0.0:3025

proxy:
  teleportConfig:
    # put your teleport.yaml proxy configuration here
    teleport:
      # The join_params section must be provided for the proxies to join the auth servers
      # By default, the chart creates a Kubernetes join token which you can use.
      join_params:
        method: kubernetes
        token_name: "RELEASE-NAME-proxy"  # replace RELEASE-NAME

>       auth_server: "RELEASE-NAME-auth.RELEASE-NAMESPACE.svc.cluster.local:3025"  # replace RELEASE-NAME and RELEASE-NAMESPACE

Author: Oh, there's a vocabulary conflict here. Helm also refers to a specific deployment of an application as a Release (because the term Deployment is coined in Kubernetes).

I'll define the RELEASE-NAME and RELEASE-NAMESPACE placeholders before the code block to remove the confusion.

I would love to have a way to mark part of the code snippets as placeholders, like what google is doing with the pink-italic-uppercase placeholders in their docs.

>       auth_server: "RELEASE-NAME-auth.RELEASE-NAMESPACE.svc.cluster.local:3025"  # replace RELEASE-NAME and RELEASE-NAMESPACE

Author: 🎉  this is exactly what I was looking for, thank you!

## Step 3/4. Setting up a Teleport cluster with Helm using a custom config

In `scratch` mode, the `teleport-cluster` Helm chart generates a minimal
configuration and lets you pass your custom configuration through the chart's values.

> `teleport-cluster` deploys two sets of nodes: proxy and auth.

Reviewer: ```suggestion
`teleport-cluster` deploys two sets of pods: proxy and auth.
```

> You must provide two configurations, one for each node type.

Reviewer: ```suggestion
You must provide two configurations, one for each pod type.
```

> - The `proxy` node configuration should contain at least the `proxy_service` section and the `teleport` section without the `storage` part.

Reviewer: ```suggestion
- The `proxy` pod configuration should contain at least the `proxy_service` section and the `teleport` section without the `storage` part.
```

> - The `auth` node configuration should contain at least the `auth_service` section and the `teleport` one.

Reviewer: ```suggestion
- The `auth` pod configuration should contain at least the `auth_service` section and the `teleport` one.
```

> For the proxy nodes to connect to the auth nodes you must provide a token in their configuration.
> The chart creates a Kubernetes join token named after the release name you can use for a seamless join.

Reviewer: ```suggestion
The chart automatically creates a Kubernetes join token named after the Helm release which will enable the proxy pods to seamlessly connect to the auth pods.
If you do not want to use this automatic token, you must provide a valid Teleport join token in the proxy pods' configuration.
```

<Admonition type="warning">
  When using `scratch` or `standalone` mode, you **must** use highly-available storage (e.g. etcd, DynamoDB, or Firestore) for multiple replicas to be supported.

  [Information on supported Teleport storage backends](../../reference/backends.mdx)

  Manually configuring NFS-based storage or `ReadWriteMany` volume claims is **NOT** supported for an HA deployment and will result in errors.
</Admonition>

Write the following `my-values.yaml` file, and adapt the teleport configuration as you want.
You can find all possible configuration fields in the [Teleport Config Reference](../../reference/config.mdx).

```yaml
chartMode: scratch

auth:
  teleportConfig:
    # put your teleport.yaml auth configuration here
    teleport:
      log:
        output: stderr
        severity: INFO

    auth_service:
      enabled: true

>       cluster_name: custom.example.com

Reviewer: ```suggestion
```

As we automatically set this to the FQDN if it isn't passed here, I recommend we don't tell people to override it. Most of our docs/guides now assume that cluster name == cluster FQDN so it's best to preserve that relationship IMO.

$ helm install teleport teleport/teleport-cluster \
  --namespace teleport \

>   --set chartMode=custom \

Author: ```suggestion
```

  It will help if you have access to the DNS provider which hosts `example.com` so you can add a `custom.example.com` record
  and point it to the external IP or hostname of the Kubernetes load balancer.

  Don't worry if you can't - you'll just have to remember to set the proxy's `public_addr` to the external IP or

>   hostname of the Kubernetes load balancer to be able to access Teleport from your local machine.

Author: I don't like this section, I'll remove it entirely.

From what I understand the remark means you can set `public_addr` to the load balancer IP address instead of a DNS entry. On such a setup, web browsers won't be able to validate the web UI and TLS multiplexing is not possible without `--insecure` all over the place.
